### PR TITLE
KFSPTS-20397 add cache to AWS Secrets manager and create testing batch job

### PR DESCRIPTION
**Pull request by jhulslander**
There is a related PR https://github.com/CU-CommunityApps/config/pull/20 that will need to be merged before the unit tests can run.

This PR adds caching to the AWS secret service, makes the secret service a prototype, so each time the service is retrieved from spring it will be given a new instance so it will have a separate cache.

This PR also adds a batch job that will we will run so we can see how well this will run from our batch infrastructure

Please wait to merge this PR till devops has a chance to update conffig files in puppet.

### CODE REVIEW

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml
Comment (cah292) [inline src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:207]: Shouldn't the "awsSecretService" bean be declared with prototype scope as well?

Also, I'm trying to figure out how the prototype setup would work out when put into actual non-testing use, as well as what the expectations are for the cache lifecycle. Would each dependent service have its own AwsSecretService instance that is only initialized at startup, and then the service's processing will call the cache management methods on its own AwsSecretService instance as needed? Or would the dependent services be expected to retrieve a new AwsSecretService instance prior to starting their next processing run? Or would the dependent services need to be declared as prototypes as well, and then the caller would have to retrieve a new instance of that service?

Example case: ConcurReportsServiceImpl is dependent on ConcurAccessTokenServiceImpl to retrieve the current access token prior to making its calls to Concur. If we update this process to use AWS secrets, ConcurAccessTokenServiceImpl would have its own AwsSecretService instance. ConcurReportsServiceImpl would then be affected by the secrets caching of ConcurAccessTokenServiceImpl, unless we update ConcurReportsServiceImpl to have its own AwsSecretService to retrieve the access token independently.

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:43]: Depending on how long each service instance or its cache is expected to live for, and on whether a given service instance could potentially be used by multiple threads, it might be better to use a ConcurrentHashMap instead of a regular HashMap. (Note, though, that ConcurrentHashMap does not allow null keys or null values.)

Also, this line could be simplified to use "<>()" instead of "<String, String>()"

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (cah292) [inline src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:24]: Another LogManager.getLogger() call that can be simplified to use the no-arg method.

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (cah292) [inline src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:121]: This line should use the Locale.US setup for the Calendar instance, like in the method above this one.

**Review by cah292 — CHANGES_REQUESTED** [review]
Changes look good in general, though I had a few suggestions for improvement, and also had some questions about how the new prototype-based setup would be put into actual use.

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:77]: good point, how about createCacheIfNotPresent

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:79]: ok

File: src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java:31]: ok

File: src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java:49]: oh, thanks noticing, I shall remove

File: src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java:46]: this is just testing and no real tokens, however, these logs weren't as helpful as I thought, I shall remove this.

File: src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationShared.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationShared.java:20]: removed

File: src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/batch/AwsSecretValidationStep.java:102]: changed to confirmInstanceSecretValues

File: src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java:1]: interesting question.  I see this batch job as something that will live for a few months and then we remove once we have confidence this works as we hope, or if we decide to go in a different direction.
I thought it would make sense to isolate this code to its own package structure so it could be removed later.  What do you think?

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:77]: Yes, a name like that sounds good.

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:43]: OK, I switched to use ConcurrentHashMap and added logic checking for null values before adding to the cache

File: src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java
Comment (cah292) [reply src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java:1]: We typically use the "org.kuali.kfs" package structure for overlays of base financials code. In my opinion, putting this AWS code in there would cause confusion as to whether it was actually an overlay of some base code, unless we documented things accordingly.

I would recommend putting this code under the "edu.cornell.kfs.sys.aws" package structure instead.

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:121]: OK

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:207]: I'll add the prototype scope to the awsSecretService bean.

I wanted to configure this such that each time Spring delivers the awsSecretService, it will be a new instance, thus creating a separate cache.  This would also provide the flexibility of each service needing to use awsSecretService, could handle the cache to its own needs.

In the case of Concur, my expectation is the following.  ConcurReportsServiceImpl and ConcurAccessTokenServiceImpl would both have awsSecretService injected with Spring.  First we go through ConcurAccessTokenServiceImpl to update the tokens.  Then we would let ConcurReportsServiceImpl perform its functions.  Both services should clear cache in awsSecretService before and after they complete processing.  Doing so would make sure both have the latest secret tokens.  Do you think there is a better approach?

File: src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/sys/aws/AmazonSecretValidationInstance.java:1]: oh!!!!!!!, thank you for noticing, I see now, they are in org, not in edu

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml
Comment (cah292) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:207]: Having each service use its own AwsSecretService instance sounds good in general. With such an approach, though, the dependent services may have to add a new method for invoking the cache-clearing method on their internal AwsSecretService instance.

For example, ConcurEventNotificationService and/or ConcurEventNotificationProcessingService will invoke methods on ConcurReportsService that make use of the access token. Once we update our Concur process to use AWS secrets, a new ConcurReportsService method for clearing the AWS cache might be needed. That way, the Concur batch step class (or one of the notification services) can clear the cache at the appropriate times.

I can't think of a better overall approach than what you had suggested. The GlobalVariables-like approach from my earlier PR ended up being too complex and not as clear when I presented it to the team.

**Comment by jhulslander** [comment]
I've made a number of changes based on the PR, this is ready for review again

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:60]: Missing space after first comma.

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:125]: Minor typo in error message.

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:85]: On a ConcurrentHashMap, a null key will cause the containsKey() and get() methods to throw an exception. You may want to check that the key is non-null before attempting to access the Map. Sorry for not clarifying that sooner.

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:95]: This message should say "provided" instead of "provide".

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:41]: This variable is not getting used at all after initialization. Is there any code that should be checking the contents or checking for a non-null object?

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:122]: Does this log statement add any value, now that the DTO's custom toString() method has been removed?

**Review by cah292 — CHANGES_REQUESTED** [review]
I found a few spots that could use some minor cleanup, but other than that the changes look good.

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:60]: fixed

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:125]: ok

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:85]: in practice, if a null secret name gets passed in, it will get translated to "kfs/local-dev/null" or something similar depending on the namespace for the environment.
However, I added some logic and throw a meaningful exception when this happens

File: src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:95]: ok

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:41]: good point, I will add a check

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:122]: replaced with the function confirming values of AmazonSecretValidationShared

**Comment by jhulslander** [comment]
I've made some additional changes based on the PR, this is ready for review again

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (cah292) [inline src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:177]: This message is not being used anywhere. Can this line be removed if the test doesn't need to do anything with the message?

Also, if this line gets removed, you could simplify the assertThrows() call to not store the returned exception in a variable.

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (cah292) [inline src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:19]: Can this import be removed? I don't see it being used anywhere in this class.

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:102]: The part to the right of the equals sign could be simplified as "new HashMap<>()" if needed.

**Review by cah292 — CHANGES_REQUESTED** [review]
I found a few more minor items to be cleaned up, but otherwise the code changes look good.

This is now at the point where we could merge the "config" PR to get the unit tests to run. I was the one who originally authored the linked PR, though. Please let me know if your test secrets are using the conventions from that PR, and if you are fine with me merging the "config" PR even though I'm its author.

**Comment by jhulslander** [comment]
@cah292 I believe the config PR is still correct.  I think it is fine if you want merge the PR,  I reviewed it earlier on.  I will work on these changes

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:88]: typo fixed

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:19]: yes

File: src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java:102]: updated

File: src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:177]: actually, I had a string comparison assertion that I must have accidentally removed, I shall put it back.

**Comment by cah292** [comment]
The related config PR has been merged. Now we should be able to see whether Jenkins can get the new unit tests to pass.

**Review by cah292 — COMMENTED** [review]
The revised changes look good, but the unit tests are still failing. Does the new Jenkins instance have a separate unit-test-related config file that needs to be updated?

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
Jenkins needed to get updated with config changes, I think I have done so manually

**Comment by jhulslander** [comment]
retest this please

**Comment by cah292** [comment]
Something is still not quite right with the configs; the tests failed again.

**Comment by jhulslander** [comment]
I don't think it is pick up the config changes ..... I noticed the job didn't clean it's workspace before starting .... I enabled that, lets see if that helps

**Comment by jhulslander** [comment]
I am not sure what is going on here ...... I will have to pick this up Monday morning

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
This PR can be merged once the unit tests pass and passes code review.  After this PR is merged, the kuali puppet PR should also be merged: https://github.com/CU-CommunityApps/kuali-puppet/pull/68 (Donivan has reviewed it and approved the PR

**Comment by jhulslander** [comment]
retest this please

**Review by cah292 — APPROVED** [review]
Changes look good, unit tests pass, and Jay has indicated that this code is ready to move forward now. Merging.

**Pull request merged**
